src/DropdownMenu/DropdownMenu.tsx

```diff
@@ -0,0 +1,158 @@
+   * `selectedItem`, `undefined` will be passed.
+   */
+  onChange?: (item?: ItemInput) => unknown
+}
+
+/**
+ * A `DropdownMenu` provides an anchor (button by default) that will open a floating menu of selectable items.  The menu can be
+ * opened and navigated using keyboard or mouse.  When an item is selected, the menu will close and the `onChange` callback will be called.
+ * If the default anchor button is used, the anchor contents will be updated with the selection.
+ */
+export function DropdownMenu({
+  renderAnchor = <T extends DropdownButtonProps>(props: T) => <DropdownButton {...props} />,
+  renderItem = Item,
+  placeholder,
+  selectedItem,
+  onChange,
+  ...listProps
+}: DropdownMenuProps): JSX.Element {
+  const anchorRef = useRef<HTMLElement>(null)
+  const [overlayRef, updateOverlayRef] = useRenderForcingRef<HTMLDivElement>()
+
+  const anchorId = `dropdownMenuAnchor-${randomId()}`
+
+  const [open, setOpen] = useState<boolean>(false)
+  const [focusType, setFocusType] = useState<null | 'anchor' | 'list'>(null)
+  const onDismiss = useCallback(() => {
+    setOpen(false)
+    setFocusType(null)
+  }, [])
+
+  const onAnchorKeyDown = useCallback(
+    (event: React.KeyboardEvent<HTMLElement>) => {
+      if (!event.defaultPrevented) {
+        if (!open) {
+          if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+            setFocusType('list')
+            setOpen(true)
+            event.preventDefault()
+          } else if (event.key === ' ' || event.key === 'Enter') {
+            setFocusType('anchor')
+            setOpen(true)
+            event.preventDefault()
+          }
+        } else if (focusType === 'anchor') {
+          if (['ArrowDown', 'ArrowUp', 'Tab', 'Enter'].indexOf(event.key) !== -1) {
+            setFocusType('list')
+            event.preventDefault()
+          } else if (event.key === 'Escape') {
+            onDismiss()
+            event.preventDefault()
+          }
+        }
+      }
+    },
+    [open, focusType, onDismiss]
+  )
+  const onAnchorClick = useCallback(
+    (event: React.MouseEvent<HTMLElement>) => {
+      if (!event.defaultPrevented && event.button === 0 && !open) {
+        setOpen(true)
+        setFocusType('anchor')
+      }
+    },
+    [open]
+  )
+
+  const {position} = useAnchoredPosition({anchorElementRef: anchorRef, floatingElementRef: overlayRef})
+
+  useFocusZone({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+  useFocusTrap({containerRef: overlayRef, disabled: !open || focusType !== 'list' || !position})
+
+  return (
+    <>
+      {renderAnchor({
+        ref: anchorRef,
+        id: anchorId,
+        'aria-labelledby': anchorId,
+        'aria-haspopup': 'listbox',
+        children: selectedItem?.text ?? placeholder,
+        onClick: onAnchorClick,
+        onKeyDown: onAnchorKeyDown
+      })}
+      {open ? (
```

Comment:
Instead of `{foo ? <SomeJSX /> : null}` it's more idiomatic to use `{foo && <SomeJSX />}`

Reply:
See https://github.com/primer/components/pull/1135#discussion_r614371832.  @colebemis linked a great article explaining why ternary is much safer to use in these situations.

Reply:
Eh, kind of a weak argument if you ask me, but fair enough!

Reply:
The tradeoff here is a ternary expression that potentially spans dozens of lines, which can get really hard to read.

Reply:
New lint rule proposal: No `&&` operators within a JSX expression such that the left side of the operator is a number or a string type :)